File: backend/transaction-service/transaction-domain/transaction-domain-core/src/main/java/com/saver/system/transaction/service/domain/entity/Transaction.java
package com.saver.system.transaction.service.domain.entity;

import com.saver.system.domain.entity.AggregateRoot;
import com.saver.system.domain.valueobject.AccountId;
import com.saver.system.domain.valueobject.Money;
import com.saver.system.domain.valueobject.TransactionId;
import com.saver.system.domain.valueobject.UserId;
import com.saver.system.transaction.service.domain.exception.TransactionDomainException;
import com.saver.system.transaction.service.domain.valueobject.TransactionAddress;
import com.saver.system.transaction.service.domain.valueobject.TransactionType;

public class Transaction extends AggregateRoot<TransactionId> {
    private final UserId userId;
    private final AccountId accountId;
    private final TransactionAddress transactionAddress;
    private final Money amount;
    protected TransactionType transactionType;

    private final String Category;

    public Transaction(Builder builder) {
        super.setId(builder.transactionId);
        this.userId = builder.userId;
        this.accountId = builder.accountId;
        this.transactionAddress = builder.transactionAddress;
        this.amount = builder.amount;
        this.Category = builder.category;
        this.transactionType = builder.transactionType;
    }

    public TransactionType getTransactionType() {
        return transactionType;
    };// need to be method

    public AccountId getAccountId() {
        return accountId;
    }

    public UserId getUserId() {
        return userId;
    }

    public Money getMoney() {
        return amount;
    }

    public TransactionAddress getTransactionAddress() {
        return transactionAddress;
    }

    public String getCategory() {
        return Category;
    }

    public void validate() {

        validateTotalAmount();
    }

    private void validateTotalAmount() {
        if (amount == null || !amount.isGreaterThanZero()) {
            throw new TransactionDomainException("money amount must be greater than zero!");
        }
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {

        private UserId userId;
        private AccountId accountId;
        private TransactionAddress transactionAddress;
        private Money amount;
        protected TransactionType transactionType;

        private TransactionId transactionId;

        private String category;

        public Builder userId(UserId userId) {
            this.userId = userId;
            return this;
        }

        public Builder accountId(AccountId accountId) {
            this.accountId = accountId;
            return this;
        }

        public Builder transactionAddress(TransactionAddress transactionAddress) {
            this.transactionAddress = transactionAddress;
            return this;
        }

        public Builder amount(Money amount) {
            this.amount = amount;
            return this;
        }

        public Builder transactionType(TransactionType transactionType) {
            this.transactionType = transactionType;
            return this;
        }

        public Builder transactionId(TransactionId transactionId) {
            this.transactionId = transactionId;
            return this;
        }

        public Builder category(String category) {
            this.category = category;
            return this;
        }

        public Transaction build() {
            return new Transaction(this);
        }

    }

}

File: backend/Services/transactions/src/domain/domain-core/entity/transaction.ts
import {
  AccountId,
  AggregateRoot,
  Money,
  TransactionStatus,
  UserId,
  TransactionAddress,
  TransactionId,
} from "@machsan-tiggo/common";
import { TransactionDomainException } from "../exception/transaction-domain-exception";

export abstract class Transaction extends AggregateRoot<TransactionId> {
  protected readonly transactionId: TransactionId;
  protected readonly userId: UserId;
  protected readonly accountId: AccountId;
  protected readonly amount: Money;
  protected readonly transactionStatus: TransactionStatus;
  protected readonly Category: string;
  protected readonly created_at: Date;
  protected readonly updated_at: Date;
  protected readonly transactionAddress: TransactionAddress;

  constructor(
    id: TransactionId,
    userId: UserId,
    accountId: AccountId,
    amount: Money,
    transactionStatus: TransactionStatus,
    Category: string,
    created_at: Date,
    updated_at: Date,
    transactionAddress: TransactionAddress
  ) {
    super(id);
    this.transactionId = id;
    this.userId = userId;
    this.accountId = accountId;
    this.amount = amount;
    this.transactionStatus = transactionStatus;
    this.Category = Category;
    this.created_at = created_at;
    this.updated_at = updated_at;
    this.transactionAddress = transactionAddress;
  }
  public validate(): void {
    this.validateTotalAmount();
  }

  private validateTotalAmount(): void {
    if (this.amount == null || !this.amount.isGreaterThanZero()) {
      throw new TransactionDomainException(
        "money amount must be greater than zero!"
      );
    }
  }

  public get TransactionId(): TransactionId {
    return this.transactionId;
  }
  getUserId(): UserId {
    return this.userId;
  }
  getAccountId(): AccountId {
    return this.accountId;
  }
  getAmount(): Money {
    return this.amount;
  }
  getTransactionStatus(): TransactionStatus {
    return this.transactionStatus;
  }
  getCategory(): string {
    return this.Category;
  }
  getCreatedAt(): Date {
    return this.created_at;
  }
  getUpdatedAt(): Date {
    return this.updated_at;
  }
  getTransactionAddress(): TransactionAddress {
    return this.transactionAddress;
  }
}
